packages/eslint-plugin/src/rules/member-ordering.ts

```diff
@@ -376,6 +384,26 @@ function getMemberName(
   }
 }
 
+/**
+ * Returns true if the member is optional based on the member type.
+ *
+ * @param node the node to be evaluated.
+ *
+ * @returns Whether the member is optional, or false if it cannot be optional at all.
+ */
+function isMemberOptional(node: Member): boolean {
+  switch (node.type) {
+    case AST_NODE_TYPES.TSPropertySignature:
+    case AST_NODE_TYPES.TSMethodSignature:
+    case AST_NODE_TYPES.TSAbstractPropertyDefinition:
+    case AST_NODE_TYPES.PropertyDefinition:
+    case AST_NODE_TYPES.TSAbstractMethodDefinition:
+    case AST_NODE_TYPES.MethodDefinition:
+      return !!node.optional;
+  }
+  return false;
```

Comment:
Indeed, codecov is right: there are some node types not tested...

```ts
type:
  | TSESTree.AST_NODE_TYPES.StaticBlock
  | TSESTree.AST_NODE_TYPES.TSCallSignatureDeclaration
  | TSESTree.AST_NODE_TYPES.TSConstructSignatureDeclaration
  | TSESTree.AST_NODE_TYPES.TSIndexSignature
```

Could you add tests that include members of these types please?

Reply:
Going off the assumption that we want all of these types to be considered as required too, how should the error message be changed for these cases? For example:
```js
      code: `
class X {
  a?: string;
  static {}
  b?: string;
}
      `,
      options: [
        {
          default: {
            memberTypes: 'never',
            order: 'as-written',
            required: 'first',
          },
        },
      ],
```
The error thrown here is `Member a should be declared after all required members.` because it considers the static block to be required. Would a user here be aware of this assumption as well?

Reply:
Hmm, that's a great point. My gut says _yes_ they should be -- and if they're not, then the rule's error existing would inform them.

The rule can already be configured to e.g. [put static fields before instance fields](https://typescript-eslint.io/rules/member-ordering/#static-fields-before-instance-fields). So I feel comfortable allowing this option in as-is.

What do you think?

Reply:
Yep, sounds good to me. We might also want the other messages to be a bit more explicit now since they can be ran within the context of required/optional members.